### PR TITLE
[SKIP CI] Add sparse build to Github Actions

### PR DESCRIPTION
**Pull request by marc-hb**
4 commits.

Main one:

The next step is to find how to extract the (too many?) errors.

In the mean time this already makes sure the build process never bitrots
and that it will always possible to use sparse. It also "documents" how
to use sparse: just copy/paste the commands run by CI.

Signed-off-by: Marc Herbert <marc.herbert@intel.com>

### CODE REVIEW

**Comment by marc-hb** [comment]
APL-only failure in https://sof-ci.01.org/sof-pr-viewer/#/build/PR6311/build10397103 is unrelated. @wszypelt could you have a look?

**Comment by wszypelt** [comment]
@marc-hb I fixed the APL and re-tested, the results should be within 20 minutes

**Review by lyakh — APPROVED** [review]
Very good, thanks a lot! As for too many errors - the only error I see in the output is the one for alsa-tools about being unable to build topology2. As for sparse errors - we did have several of them in the past - both in Zephyr and in SOF, but we fixed them early when adding sparse support. Now I see only warnings left and yes, we should gradually fix them - now that we'll have sparse in CI. After merging this, to make sparse even more useful, can we make it fail in two cases:
1. if any "error:" messages from sparse are detected
2. if any "address space" sparse warnings are detected

File: .github/workflows/zephyr.yml
Comment (aborisovich) [inline .github/workflows/zephyr.yml:79]: > The sparse build has the worst error handling ever.

So we need to improve it, not build even worst solutions on top of something bad.
We are already very bad on projects integrations, lets not degrade further if possible.
Please go, submit a PR to Zephyr project, add `find_program` for sparse executable, then error message displayed will have proper content (that sparse executable had not been found).

**Review by aborisovich — CHANGES_REQUESTED** [review]
Overall this is a no-go for me in terms of integration.
We should make it right by modifications in Zephyr, see my comments.

File: zephyr/docker-build.sh
Comment (aborisovich) [inline zephyr/docker-build.sh:21]: Please add following code to Zephyr project, file:
(zephyr)/cmake/compiler/gcc/target.cmake:
image

**Comment by marc-hb** [comment]
@aborisovich I can and I will spend some (limited) amount of time trying to fix error handling in the sparse build in Zephyr (which I did not implement myself) but...  I don't see what difference it will make to this PR besides reducing the size of the comments. Could you elaborate?

Moreover Zephyr is in pre-release freeze so I hope I can keep said fix small and harmless looking...

> So we need to improve it, not build even worst solutions on top of something bad.
> We are already very bad on projects integrations, lets not degrade further if possible

I don't really see what this PR is "making worse"... it's just setting a couple of variables in the right places, running already existing scripts and adding one basic Github Action. It was very hard to find the right "magic" because of the poor error handling but that magic ended up being pretty basic and small when the user (=me) does not fumble anything.

File: zephyr/docker-build.sh
Comment (marc-hb) [reply zephyr/docker-build.sh:21]: Thx for the direct link

**Review by marc-hb — COMMENTED** [review]
Zephyr error handling fixed, ugly comments removed.

Replaced with an explanation of why sparse exit status sucks and why this check is "always green" for now.

I could not make any other improvement.

**Comment by lgirdwood** [comment]
@aborisovich good for you now ?

**Review by lyakh — APPROVED** [review]
I'd actually prefer to have a complete solution directly - i.e. one that actually *shows* sparse regressions as CI *failures* and not just runs the test and always succeeds. We can take this PR as the first step but we actually need a working solution ASAP, we need it to do useful work and not just to exercise a neat feature

**Comment by marc-hb** [comment]
> We can take this PR as the first step but we actually need a working solution ASAP, 

Agreed but note sparse's inability to provide a useful exit status is a sparse problem, not a CI problem. It affects any build system anywhere in CI and not.

**Review by aborisovich — APPROVED** [review]
Looks as good as it can be for now, thanks @marc-hb! 👍 
Missing error codes, parsing output by regexes, where did I saw it last time... Ah! Docker.

**Comment by aborisovich** [comment]
> @aborisovich good for you now ?

Sorry for late reply, I had been on a sick leave.

**Review by lgirdwood — APPROVED** [review]

**Comment by lyakh** [comment]
@marc-hb any ETA for part two - parsing sparse log and detecting "address space" errors? And I think we need '--cmake-args=-DEXTRA_CFLAGS=-Wsparse-error' too to make sparse return an error for detected errors. BTW, with that the compilation won't be stopped, or will it? If yes, then maybe better parse output for "error:" strings too instead of stopping compilation

**Comment by lgirdwood** [comment]
@marc-hb we do need part 2 to be enabled, IIUC it wont be perfect but it's better than what we have today.

**Comment by marc-hb** [comment]
@lgirdwood part 2 is not a CI problem at all, it's a pure sparse issue. developers need the exact same sparse fix or workaround.

**Comment by lyakh** [comment]
@marc-hb sorry, not sure I understand. Developers might or might not run sparse manually, same as they sometimes do and sometimes don't build-test for different platforms, different configurations, etc. If we do run sparse locally - sure, we'd grep output for error strings - the same ones that the GitHub hook now should grep for.

**Comment by marc-hb** [comment]
> sure, we'd grep output for error strings - the same ones that the GitHub hook now should grep for.

Github searches for the same errors as developers, not the other way around. Github merely does what developers do; Github merely runs developers' scripts, not the other way round (except for Docker scripts which are rarely desired outside CI)

Github does not know anything about sparse, it's a blackbox - garbage in, garbage out. Only (some) developers know how to use and configure sparse, what sparse errors to look for, which sparse flags to use, how to work around sparse bugs and limitations, etc.. Anyone who understands sparse besides you? I don't.

**Comment by lgirdwood** [comment]
@marc-hb I dont think you need to understand sparse for this initial integration. I agree, we cant immediately "run" here, but we can "walk". We need to have an option in the SOF build script (that is used by CI today) to do something like
```sh
make target SPARSE=1 | grep "bad" > output.file
# output file should be empty for a pass
if (output.file > 0 bytes) 
    emit CI error and dump output.file
```
We have very complex usages around cache and need to give the maintainers some way of knowing whether a PR is good.
@lyakh has told me we only need to grep for one string today. We can always update and improve as we go on, but we do need something like this today.

**Comment by marc-hb** [comment]
Done:
- #6510

**Pull request merged**
